agent/connect_auth.go

```diff
@@ -0,0 +1,146 @@
+package agent
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/cache"
+	cachetype "github.com/hashicorp/consul/agent/cache-types"
+	"github.com/hashicorp/consul/agent/connect"
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+// ConnectAuthorize implements the core authorization logic for Connect. It's in
+// a separate agent method here because we need to re-use this both in our own
+// HTTP API authz endpoint and in the gRPX xDS/ext_authz API for envoy.
+//
+// The ACL token and the auth request are provided and the auth decision (true
+// means authorised) and reason string are returned.
+//
+// If the request input is invalide the error returned will be a
+// BadRequestError, if the token doesn't grant necessary access then an
+// acl.ErrPermissionDenied error is returned, otherwise error indicates an
+// unexpected server failure. If access is denied, no error is returned but the
+// first return value is false.
+func (a *Agent) ConnectAuthorize(token string,
+	req *structs.ConnectAuthorizeRequest) (authz bool, reason string, m *cache.ResultMeta, err error) {
+	if req == nil {
+		err = BadRequestError{"Invalid request"}
+		return
```

Comment:
I think this method is long enough you shouldn't rely on the named parameters shortcut to return. Below there is at least one place that `err` is shadowed and then for the rest of the method body those errors go into a pit of sadness.

Reply:
Yeah I dislike named return ergonmics generally and tend to not use them. The value I saw here was in naming/expressing intent about the return args which seems nice given how many there are.

I'll refactor this to use a helper or something to make it less gross to read and reason about.